compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -191,6 +191,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         rcvr_opt: Option<&'tcx hir::Expr<'tcx>>,
         rcvr_ty: Ty<'tcx>,
         item_name: Ident,
+        expr_id: hir::HirId,
```

Comment:
And the span too

Reply:
I looked into it. Not much is derivable from `expr_id`. 

These are the two calls to `report_method_error`: https://github.com/rust-lang/rust/blob/9cdfe285ca724c801dc9f78d22b24ea69b787f26/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs#L835-L845 and https://github.com/rust-lang/rust/blob/9cdfe285ca724c801dc9f78d22b24ea69b787f26/compiler/rustc_hir_typeck/src/expr.rs#L1347-L1357

As you can see `rcvr_opt` (second arg) is being explicitly passed as `None` in one case and `Some(rcvr)` in the other. Similarly `rcvr_ty` (third arg) is `ty.normalized` in one case and `rcvr_ty` in the other and both are obtained in different ways. `source` (fifth arg) too has different values with somewhat convoluted provenance and the same can be said of `args` (seventh arg).

So I don't feel we should change any of this.

Reply:
`SelfSource`, `rcvr_opt`, and `args` are all a property of whether the `expr_id` (which really should be passed in as a whole `hir::Expr`) is `ExprKind::MethodCall` or `ExprKind::Path`.

Reply:
Similarly, `segment.ident` should be derivable from that expression (it's directly stored in the `MethodCall` and is also as the final segment in `ExprKind::Path`, or something like that).

I don't think I said to touch `rcvr_ty`, which can continue to be passed in.

Reply:
Similarly, `span` can be derived from the expression without much trouble as well.

Reply:
Okay, let me have another look at it

Reply:
Turns out the `expr_id` isn't necessarily pointing to an `Expr` always. it could also be  a `Pat` because `check_pat` also calls `report_method_error` (`check_pat`->`resolve_ty_and_res_fully_qualified_call`->`report_method_error`) . So if I were to derive all these bits from `expr_id` I'll have to take `Pat` into account in addition to `Expr`, which means I might have to repeat code like this from `resolve_ty_and_res_fully_qualified_call`: https://github.com/rust-lang/rust/blob/78dd504f2fd87c0cfabff7d9174253411caf2f80/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs#L749-L772
within `report_method_error`. I have a feeling that complicates the code instead of making it simpler 😫

Reply:
That didn't end up being too complicated, for the record: https://github.com/rust-lang/rust/pull/125906/commits/2227c1fb5ce11961a98fd7d67d21be60bcd1d42e